src/prefect_server/api/flows.py

```diff
@@ -261,37 +267,57 @@ async def create_flow(
         is_schedule_active=False,
-        tasks=[
-            models.Task(
-                id=t.id,
-                tenant_id=tenant_id,
-                name=t.name,
-                slug=t.slug,
-                type=t.type,
-                max_retries=t.max_retries,
-                tags=t.tags,
-                retry_delay=t.retry_delay,
-                trigger=t.trigger,
-                mapped=t.mapped,
-                auto_generated=t.auto_generated,
-                cache_key=t.cache_key,
-                is_reference_task=t.is_reference_task,
-                is_root_task=t.is_root_task,
-                is_terminal_task=t.is_terminal_task,
+        tasks=[],
+        edges=[],
+    ).insert()
+
+    try:
+        batch_insertion_size = 2500
+
+        for tasks_chunk in chunked_iterable(flow.tasks, batch_insertion_size):
+            await models.Task.insert_many(
+                [
+                    models.Task(
+                        id=t.id,
+                        flow_id=flow_id,
+                        tenant_id=tenant_id,
+                        name=t.name,
+                        slug=t.slug,
+                        type=t.type,
+                        max_retries=t.max_retries,
+                        tags=t.tags,
+                        retry_delay=t.retry_delay,
+                        trigger=t.trigger,
+                        mapped=t.mapped,
+                        auto_generated=t.auto_generated,
+                        cache_key=t.cache_key,
+                        is_reference_task=t.is_reference_task,
+                        is_root_task=t.is_root_task,
+                        is_terminal_task=t.is_terminal_task,
+                    )
+                    for t in tasks_chunk
+                ]
             )
-            for t in flow.tasks
-        ],
-        edges=[
-            models.Edge(
-                tenant_id=tenant_id,
-                upstream_task_id=task_lookup[e.upstream_task].id,
-                downstream_task_id=task_lookup[e.downstream_task].id,
-                key=e.key,
-                mapped=e.mapped,
+
+        for edges_chunk in chunked_iterable(flow.edges, batch_insertion_size):
+            await models.Edge.insert_many(
+                [
+                    models.Edge(
+                        tenant_id=tenant_id,
+                        flow_id=flow_id,
+                        upstream_task_id=task_lookup[e.upstream_task].id,
+                        downstream_task_id=task_lookup[e.downstream_task].id,
+                        key=e.key,
+                        mapped=e.mapped,
+                    )
+                    for e in edges_chunk
+                ]
             )
-            for e in flow.edges
-        ],
-    ).insert()
+
+    except Exception as exception:
+        logger.error(exception)
+        await api.flows.delete_flow(flow_id=flow_id)
+        raise Exception("Failed to register Flow") from exception
```

Comment:
I would do this

```suggestion
    except Exception as exc:
        logger.error("`create_flow` failed during insertion", exc_info=True)
        await api.flows.delete_flow(flow_id=flow_id)
        raise APIError() from exc
```

- The log now explains what happens and attaches a full traceback
- The `APIError` will hide all further exception info from the user but still be chained to the original exception for tracking in GCP

Reply:
love it